### PR TITLE
Move convention extensions from RazorPagesOptions to PageConventionCo…

### PR DESCRIPTION
**Pull request by pranavkm**
…llectionsExtensions

### CODE REVIEW

**Comment by pranavkm** [comment]
I'll update the docs once we settle on the design.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+namespace Microsoft.AspNetCore.Mvc.ApplicationModels
+{
+    public class PageConventionCollection : Collection<IPageConvention>
+    {
+        public IPageApplicationModelConvention AddConvention(string pageName, Action<PageApplicationModel> action)
+        {
+            if (string.IsNullOrEmpty(pageName))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(pageName));
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: Should we check for a leading slash here?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+using System;
+using System.Collections.ObjectModel;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationModels
+{
+    public class PageConventionCollection : Collection<IPageConvention>
+    {
+        public IPageApplicationModelConvention AddConvention(string pageName, Action<PageApplicationModel> action)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: This kind of overloading `Action<PageApplicationModel> action` vs `Action<PageRouteModel> action` is not good. These need to have different names

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+        /// <summary>
+        /// Creates and adds an <see cref="IPageRouteModelConvention"/> that invokes an action on
+        /// <see cref="PageRouteModel"/> instances for all page under the specified folder.
+        /// </summary>
+        /// <param name="folderPath">The path of the folder relative to the Razor Pages root. e.g. <c>/Users/</c></param>
+        /// <param name="action">The <see cref="Action"/>.</param>
+        /// <returns>The added <see cref="IPageApplicationModelConvention"/>.</returns>
+        public IPageRouteModelConvention AddFolderConvention(string folderPath, Action<PageRouteModel> action)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: AddFolderRouteModelConvention

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            {
+                throw new ArgumentException(Resources.PathMustBeRootRelativePath, nameof(folderPath));
+            }
+        }
+
+        private TConvention Add<TConvention>(TConvention convention) where TConvention: IPageConvention
+        {
+            base.Add(convention);
+            return convention;
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: lolwut

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageConventionCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,221 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Microsoft.AspNetCore.Mvc.Authorization;
+using Microsoft.AspNetCore.Mvc.Filters;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageConventionCollectionExtensions.cs]: Should this be in the DI namespace?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Resources.resx [hunk 6]
```diff
@@ -150,4 +150,7 @@
   <data name="InvalidPageType_NoModelProperty" xml:space="preserve">
     <value>The type '{0}' is not a valid page. A page must define a public, non-static '{1}' property.</value>
   </data>
+  <data name="InvalidValidPageName" xml:space="preserve">
+    <value>'{0}' is not a valid page name. A page name is path relative to the Razor Pages root directory that starts with a leading forward slasg ('/') and does not contain the file extension.</value>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Resources.resx]: slasg

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Resources.resx]: Can you add an example

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/PageConventionCollectionExtensionsTest.cs [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        private static void ApplyConventions(PageConventionCollection conventions, PageApplicationModel[] models)
+        {
+            foreach (var convention in conventions.OfType<IPageApplicationModelConvention>())
+            {
+                foreach (var model in models)
+                {
+                    convention.Apply(model);
+                }
+            }
+        }
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/PageConventionCollectionExtensionsTest.cs]: This is a really extreme amount of tests for something simple

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageConventionCollection.cs]: Saves me 8 lines of code!

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageConventionCollectionExtensions.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageConventionCollectionExtensions.cs]: Sure, It's one fewer ns to import in your Startup.

**Pull request closed without merging**